worker/src/scripts/upsertDefaultModelPrices.ts

```diff
@@ -0,0 +1,192 @@
+import { z } from "zod";
+import { prisma } from "@langfuse/shared/src/db";
+import defaultModelPrices from "../constants/default-model-prices.json";
+import { logger } from "@langfuse/shared/src/server";
+
+const DefaultModelPriceSchema = z.object({
+  id: z.string(),
+  model_name: z.string(),
+  match_pattern: z.string(),
+  created_at: z.coerce.date(),
+  updated_at: z.coerce.date(),
+  prices: z.record(z.number()),
+  tokenizer_config: z.record(z.union([z.string(), z.number()])).nullish(),
+  tokenizer_id: z.string().nullish(),
+});
+
+/**
+ * Upserts default model prices into the database into models and prices tables.
+ *
+ * This function performs the following operations:
+ * 1. Fetches existing default models from the database (single query, not in transaction).
+ * 2. Parses and validates the default model prices from the JSON file in the constants folder.
+ * 3. Processes the default model prices in batches.
+ *
+ * Transaction behavior:
+ * - Each batch is processed in parallel
+ * - Within a batch, each model upsert and corresponding price upsert are in the same transaction
+ *
+ * Batching:
+ * - Default model prices are processed in batches of 10 to optimize performance / not overwhelm the database
+ *
+ * Server start-time overhead:
+ * - If all models are up-to-date and 'force' is false, only the initial query to fetch
+ *   existing model update dates will be executed.
+ *
+ * @param force - If true, updates all models regardless of their last update time.
+ *                If false, only updates models that are outdated.
+ */
+
+export const upsertDefaultModelPrices = async (force = false) => {
+  const startTime = Date.now();
+  try {
+    logger.debug(`Starting upsert of default model prices (force = ${force})`);
+
+    const parsedDefaultModelPrices = z
+      .array(DefaultModelPriceSchema)
+      .parse(defaultModelPrices);
+
+    // Fetch existing default models. Store in a map for O(1) lookup.
+    const existingModelUpdateDates = new Map(
+      (
+        await prisma.model.findMany({
+          where: {
+            projectId: null,
+          },
+          select: {
+            id: true,
+            updatedAt: true,
+          },
+        })
+      ).map((model) => [model.id, model.updatedAt])
+    );
```

Comment:
Do you see any issues in doing this part here outside of the update transaction? If multiple workers start at the same time, which would be expected during updates, all of them may read the same initial state and then make updates based on outdated knowledge.

Reply:
Similar line of thinking as above - even if multiple workers attempt the same updates, they are quite cheap and were not yet worth heavily optimizing. Let me know if you see a clear need for that?